src/protocols/HyprlandSurface.cpp

```diff
@@ -0,0 +1,113 @@
+#include "HyprlandSurface.hpp"
+#include "../desktop/WLSurface.hpp"
+#include "../render/Renderer.hpp"
+#include "core/Compositor.hpp"
+#include "hyprland-surface-v1.hpp"
+
+CHyprlandSurface::CHyprlandSurface(SP<CHyprlandSurfaceV1> resource, SP<CWLSurfaceResource> surface) : m_pSurface(surface) {
+    setResource(std::move(resource));
+}
+
+bool CHyprlandSurface::good() const {
+    return m_pResource->resource();
+}
+
+void CHyprlandSurface::setResource(SP<CHyprlandSurfaceV1> resource) {
+    m_pResource = std::move(resource);
+
+    if (!m_pResource->resource())
+        return;
+
+    m_pResource->setDestroy([this](CHyprlandSurfaceV1* resource) { destroy(); });
+    m_pResource->setOnDestroy([this](CHyprlandSurfaceV1* resource) { destroy(); });
+
+    m_pResource->setSetOpacity([this](CHyprlandSurfaceV1* resource, uint32_t opacity) {
+        if (!m_pSurface) {
+            m_pResource->error(HYPRLAND_SURFACE_V1_ERROR_NO_SURFACE, "set_opacity called for destroyed wl_surface");
+            return;
+        }
+
+        auto fOpacity = wl_fixed_to_double(opacity);
+        if (fOpacity < 0.0 || fOpacity > 1.0) {
+            m_pResource->error(HYPRLAND_SURFACE_V1_ERROR_OUT_OF_RANGE, "set_opacity called with an opacity value larger than 1.0 or smaller than 0.0.");
+            return;
+        }
+
+        m_fOpacity = fOpacity;
+    });
+
+    listeners.surfaceCommitted = m_pSurface->events.commit.registerListener([this](std::any data) {
+        auto surface = CWLSurface::fromResource(m_pSurface.lock());
+
+        if (surface && surface->m_fOverallOpacity != m_fOpacity) {
+            surface->m_fOverallOpacity = m_fOpacity;
+            auto box                   = surface->getSurfaceBoxGlobal();
+
+            if (box.has_value())
+                g_pHyprRenderer->damageBox(&*box);
+
+            if (!m_pResource)
+                PROTO::hyprlandSurface->destroySurface(this);
+        }
+    });
+
+    listeners.surfaceDestroyed = m_pSurface->events.destroy.registerListener([this](std::any data) {
+        m_pSurface.reset();
```

Comment:
useless

Reply:
Since its a WP?

Reply:
surface calls destroy when its about to be freed so the WP will be dead after this is called

Reply:
fixed